src/NonLinearProgram/nlp_utilities.jl

```diff
@@ -0,0 +1,556 @@
+        V_L[i, j] = _V_L[j]
+        X_lb[i, i] = _X[j] - _X_L[j]
+        I_L[j, i] = -1
+    end
+    # value and dual of the upper bounds
+    for (i, j) in enumerate(has_up)
+        V_U[i, j] = _V_U[j]
+        X_ub[i, i] = _X_U[j] - _X[j]
+        I_U[j, i] = 1
+    end
+
+    # Function Derivatives
+    hessian, jacobian = compute_optimal_hess_jac(model, cons)
+    primal_idx = [i.value for i in model.cache.primal_vars]
+    params_idx = [i.value for i in model.cache.params]
+    # Hessian of the lagrangian wrt the primal variables
+    W = spzeros(num_vars + num_ineq, num_vars + num_ineq)
+    W[1:num_vars, 1:num_vars] = hessian[primal_idx, primal_idx]
+    # Jacobian of the constraints
+    A = spzeros(num_cons, num_vars + num_ineq)
+    # A is the Jacobian of: c(x) = b and c(x) <= b and c(x) >= b, possibly all mixed up.
+    # Each of the will be re-written as:
+    # c(x) - b = 0
+    # c(x) - b - su = 0, su <= 0
+    # c(x) - b - sl = 0, sl >= 0
+    # Jacobian of the constraints wrt the primal variables
+    A[:, 1:num_vars] = jacobian[:, primal_idx]
+    # Jacobian of the constraints wrt the slack variables
+    for (i, j) in enumerate(geq_locations)
+        A[j, num_vars+i] = -1
+    end
+    for (i, j) in enumerate(leq_locations)
+        A[j, num_vars+i] = -1
+    end
+    # Partial second derivative of the lagrangian wrt primal solution and parameters
+    ∇ₓₚL = spzeros(num_vars + num_ineq, num_parms)
+    ∇ₓₚL[1:num_vars, :] = hessian[primal_idx, params_idx]
+    # Partial derivative of the equality constraintswith wrt parameters
+    ∇ₚC = jacobian[:, params_idx]
+
+    # M matrix
+    # M = [
+    #     [W A' -I I];
+    #     [A 0 0 0];
+    #     [V_L 0 (X - X_L) 0]
+    #     [V_U 0 0 0 (X_U - X)]
+    # ]
+    len_w = num_vars + num_ineq
+    M = spzeros(
+        len_w + num_cons + num_low + num_up,
+        len_w + num_cons + num_low + num_up,
+    )
+
+    M[1:len_w, 1:len_w] = W
+    M[1:len_w, len_w+1:len_w+num_cons] = A'
+    M[len_w+1:len_w+num_cons, 1:len_w] = A
+    M[1:len_w, len_w+num_cons+1:len_w+num_cons+num_low] = I_L
+    M[len_w+num_cons+1:len_w+num_cons+num_low, 1:len_w] = V_L
+    M[
+        len_w+num_cons+1:len_w+num_cons+num_low,
+        len_w+num_cons+1:len_w+num_cons+num_low,
+    ] = X_lb
+    M[len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up, 1:len_w] = V_U
+    M[
+        len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up,
+        len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up,
+    ] = X_ub
+    M[1:len_w, len_w+num_cons+num_low+1:end] = I_U
+
+    # N matrix
+    # N = [∇ₓₚL ; ∇ₚC; zeros(num_low + num_up, num_parms)]
+    N = spzeros(len_w + num_cons + num_low + num_up, num_parms)
+    N[1:len_w, :] = ∇ₓₚL
+    N[len_w+1:len_w+num_cons, :] = ∇ₚC
+
+    return M, N
+end
+
+"""
+    inertia_corrector_factorization(M::SparseMatrixCSC, num_w, num_cons; st=1e-6, max_corrections=50)
+
+Inertia correction for the factorization of the KKT matrix. Sparse version.
```

Comment:
The theoretical question is: do we need to run inertia correction at the optimum? It could be necessary if SOSC is not satisfied at the solution indeed, but in that case I am not sure the sIpopt's formula remains valid.

Reply:
I think you are right, but couldn't there be numerical errors? 

Moreover, I believe the rest of DiffOpt just assumes that the necessary conditions hold (even if they don't). The thing is that their method doesn't fail when it doesn't hold, while sIpopt method appears to throw singular matrix error when either the conditions don't hold or when we have numerical errors. Therefore, I tried to make it consistent.

Reply:
Maybe we should log a warning if inertia correction is necessary (apologies if you're already doing this) or add an option `allow_intertia_correction` that defaults to true?

Reply:
done!

Reply:
inertial correction should be set as a MOI attribute, not as a keyword